File: asm/inout.c
// TODO: Port these to real assembly, instead of single line
// C functions.
#include <stdint.h>

unsigned char inb (unsigned short _port)
{
    unsigned char rv;
    __asm__ __volatile__ ("inb %1, %0" : "=a" (rv) : "dN" (_port));
    return rv;
}

void outb (unsigned short _port, unsigned char _data)
{
    __asm__ __volatile__ ("outb %1, %0" : : "dN" (_port), "a" (_data));
}

uint32_t inl(uint16_t _port)
{
    uint32_t rv;
    __asm__ __volatile__ ("inl %1, %0" : "=a" (rv) : "dN" (_port));
    return rv;
}
unsigned long inw (unsigned short _port)
{
    uint16_t rv;
    __asm__ __volatile__ ("inw %1, %0" : "=a" (rv) : "dN" (_port));
    return rv;
}

void outl (uint16_t _port, uint32_t _data)
{
    __asm__ __volatile__ ("outl %1, %0" : : "dN" (_port), "a" (_data));
}

File: asm/inout.go
// Package asm provides wrappers around assembly instructions to be
// called with Go syntax.
package asm

// INB executes an intel INB assembly instruction. Returns a byte from
// the input device at port.
func INB(port uint16) byte {
	// for some reason if this isn't a separate call, GCCGO strips it
	// out of the object file entirely. It's probably something I'm
	// missing.
	return inb(port)
}

//extern inb
func inb(port uint16) byte

// OUTB executes the Intel OUTB assembly instruction. Sends a byte to
// the output device at port.
func OUTB(port uint16, data byte) {
	outb(port, data)
}

//
//extern outb
func outb(port uint16, data byte)

// INL executes the Intel INL assembly instruction. Returns a uint32
// from the output device at port.
func INL(port uint16) uint32 {
	return inl(port)
}

//extern inl
func inl(port uint16) uint32

// OUTL executes the Intel OUTL assembly instruction. Writes a uint32
// to the output device at port.
func OUTL(port uint16, data uint32) {
	outl(port, data)
}

//extern outl
func outl(port uint16, data uint32)
